source/Engine.cpp

```diff
@@ -2288,21 +2297,40 @@ void Engine::DoCollection(Flotsam &flotsam)
 			break;
 		}
 	}
+	// If the flotsam was not collected, give tractor beam systems a chance to
+	// pull it.
 	if(!collector)
+	{
+		// Keep track of the origin and strength of pull that each tractor beam
+		// has on this flotsam.
+		vector<pair<Point, double>> tractorBeams;
+		for(Ship *ship : hasTractorBeam)
+			ship->FireTractorBeam(flotsam, visuals, tractorBeams);
+
+		// Find the net effect of all the tractor beams pulling on
+		// this flotsam.
+		Point pullVector;
+		for(auto &weapon : tractorBeams)
+		{
+			Point direction = (weapon.first - flotsam.Position()).Unit();
+			double magnitude = weapon.second;
+			pullVector += direction * magnitude;
+		}
```

Comment:
why not do this in one go? I dont see why you have to do two loops

Reply:
No real reason to have it be two separate loops instead of a nested one. This is just the way I thought of doing it first for whatever reason.